File: Java/122BestTimeToBuyAndSellStock.java
/* We really want to be able to do this in one parse of the array, so we keep track of the current buy price (the min) and the maximum possible profit. */

public class 122BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        int buyPrice = prices[0];
        int maxProfit = 0;

        for (int i = 1; i < prices.length; i++) {
            if (prices[i] < buyPrice) {
                buyPrice = prices[i];
            } else if (prices[i] > buyPrice) {
                maxProfit = Math.max(maxProfit, prices[i] - buyPrice);
            }
        }

        return maxProfit;
    }
}

File: Python/122BestTimeToBuyAndSellStock.py
# We only want to parse through the list once, so we keep track of the current buy price (the current minimum) and also the maximum profit we see so far.

class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        buy = prices[0]
        profit = 0

        for p in prices:
            if p < buy:
                buy = p
            else:
                profit = max(profit, p - buy)

        return profit